src/main/java/com/meta/chatbridge/llm/OpenAIPlugin.java

```diff
@@ -0,0 +1,191 @@
+      // This is honestly a guess, it's undocumented
+      functionTokens = tokenEncoding.countTokens(MAPPER.writeValueAsString(functions));
+    }
+
+    ArrayNode output = MAPPER.createArrayNode();
+    int totalTokens = functionTokens;
+    totalTokens += 3; // every reply is primed with <|start|>assistant<|message|>
+
+    JsonNode systemMessage = messages.get(0);
+    boolean hasSystemMessage = systemMessage.get("role").textValue().equals("system");
+    if (hasSystemMessage) {
+      // if the system message is present it's required
+      totalTokens += tokenCount(messages.get(0));
+    }
+    for (int i = messages.size() - 1; i >= 0; i--) {
+      JsonNode m = messages.get(i);
+      String role = m.get("role").textValue();
+      if (role.equals("system")) {
+        continue; // system has already been counted
+      }
+      totalTokens += tokenCount(m);
+      if (totalTokens > config.maxInputTokens()) {
+        break;
+      }
+      output.insert(0, m);
+    }
+    if (hasSystemMessage) {
+      output.insert(0, systemMessage);
+    }
+
+    if ((hasSystemMessage && output.size() <= 1) || output.isEmpty()) {
+      return Optional.empty();
+    }
+
+    return Optional.of(output);
+  }
+
+  @Override
+  public T handle(MessageStack<T> messageStack) throws IOException {
+    T fromUser = messageStack.tail();
+
+    ObjectNode body = MAPPER.createObjectNode();
+    body.put("model", config.model().properties().name())
+        // .put("function_call", "auto") // Update when we support functions
+        .put("n", 1)
+        .put("stream", false)
+        .put("user", fromUser.senderId().toString());
+    config.topP().ifPresent(v -> body.put("top_p", v));
+    config.temperature().ifPresent(v -> body.put("temperature", v));
+    config.maxOutputTokens().ifPresent(v -> body.put("max_tokens", v));
+    config.presencePenalty().ifPresent(v -> body.put("presence_penalty", v));
+    config.frequencyPenalty().ifPresent(v -> body.put("frequency_penalty", v));
+    if (!config.logitBias().isEmpty()) {
+      body.set("logit_bias", MAPPER.valueToTree(config.logitBias()));
+    }
+    if (!config.stop().isEmpty()) {
+      body.set("stop", MAPPER.valueToTree(config.stop()));
+    }
+
+    ArrayNode messages = MAPPER.createArrayNode();
+    config
+        .systemMessage()
+        .ifPresent(
+            m ->
+                messages
+                    .addObject()
+                    .put("role", Role.SYSTEM.toString().toLowerCase())
+                    .put("content", m));
+    for (T message : messageStack.messages()) {
+      messages
+          .addObject()
+          .put("role", message.role().toString().toLowerCase())
+          .put("content", message.message());
+    }
+
+    Optional<ArrayNode> prunedMessages = pruneMessages(messages, null);
+    if (prunedMessages.isEmpty()) {
+      return messageStack.newMessageFromBot(
+          Instant.now(), "I'm sorry but that request was too long for me.");
```

Comment:
If this is getting send back to the user, maybe add something to ask for them to repeat it in fewer words?

Reply:
Yea I thought about that and I want to get some feedback from users here, maybe make it customizable, I'm not sure.